packages/flutter_tools/lib/src/dart/pub.dart

```diff
@@ -48,19 +48,43 @@ void joinCaches({
     final String newPath = fileSystem.path.join(targetDirectory.path, entity.basename);
     if (entity is File) {
       final File file = fileSystem.file(entity.path);
-      file.renameSync(newPath);
+      file.copySync(newPath);
+      file.deleteSync(recursive: true);
     }
     else if (entity is Directory) {
       final Directory currentDirectory = fileSystem.directory(entity.path);
-      try {
-        currentDirectory.renameSync(newPath);
-      }
-      on FileSystemException { // directory already exists
-        joinCaches(fileSystem: fileSystem, targetPath: newPath, extraPath: currentDirectory.path);
+      if (!currentDirectory.existsSync()) {
+        final Directory newDirectory = targetDirectory.childDirectory(entity.basename);
+        newDirectory.createSync();
+        joinCaches(
+            fileSystem: fileSystem,
+            targetPath: newDirectory.path,
+            extraPath: currentDirectory.path
+        );
       }
     }
   }
-  extraDirectory.deleteSync();
+}
+
+/// Valid pubCache should look like this ./localCachePath/.pub-cache/hosted/pub.dartlang.org
+/// In order to join both cache both local and global should should be present in the users environment
+bool needsToJoinCache({
+  required FileSystem fileSystem,
+  required String localCachePath,
+  required String globalCachePath
+}) {
+  final Directory globalDirectory = fileSystem.directory(fileSystem.path.join(globalCachePath, '.pub-cache'));
+  final Directory localDirectory = fileSystem.directory(fileSystem.path.join(localCachePath, '.pub-cache'));
+
+  if (globalDirectory.existsSync() && localDirectory.existsSync()) {
```

Comment:
I think you can make this code by skipping this check, and instead going directly down to `pub.dartlang.org` dir and checking that those exist.